### PR TITLE
feat(learn): Add tests for refactored American/British Translator

### PR DESCRIPTION
**Pull request by SaintPeter**
Checklist:
- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

@nhcarrigan and I were at it again, updating the tests to work with the refactored project.  These have been tested against my version of the refactored project.

Here is the issue reference:
https://github.com/freeCodeCamp/freeCodeCamp/issues/39692

I have a live version of the project here:
https://american-british-translator.pearlygatesoftware.com/

### CODE REVIEW

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/american-british-translator.md [hunk 1]
```diff
@@ -11,11 +11,21 @@ Build a full stack JavaScript app that is functionally similar to this: <a href=
...
 ## Instructions
 <section id='instructions'>
-
+<ul>
+<li>All logic can go into <code>/components/translator.js</code>.</li>
+<li>Create all of the unit/functional tests in <code>tests/1_unit-tests.js</code> and <code>tests/2_functional-tests.js</code>.</li>
+<li>See the JavaScript files in <code>/components</code> for the different spelling and terms your application should translate</li>
+<li>To run the tests on Repl.it, set NODE_ENV to test without quotes in the .env file.</li>
+<li>To run the tests in the console, use the command npm run test. To open the Repl.it console, press Ctrl+Shift+P (Cmd if on a Mac) and type "open shell".</li>
+</ul>
```
Comment (Sky020) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/american-british-translator.md:28]:
```suggestion

- All logic can go into `/components/translator.js`
- Create all of the unit/functional tests in `tests/1_unit-tests.js` and `tests/2_functional-tests.js`
- See the JavaScript files in `/components` for the different spelling and terms your application should translate
- To run the tests on Repl.it, set NODE_ENV to test without quotes in the .env file
- To run the tests in the console, use the command npm run test. To open the Repl.it console, press Ctrl+Shift+P (Cmd if on a Mac) and type "open shell"

```

**Review by ShaunSHamilton — DISMISSED** [review]
Thanks, for all the hard work on this Rex and Nich.

These changes are:
1) In an effort to move to more MD syntax
2) I find the HTML lists alter the font-face on the learn platform

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/american-british-translator.md [hunk 2]
```diff
@@ -11,11 +11,21 @@ Build a full stack JavaScript app that is functionally similar to this: <a href=
 Working on this project will involve you writing your code on Repl.it on our starter project. After completing this project you can copy your public Repl.it URL (to the homepage of your app) into this screen to test it! Optionally you may choose to write your project on another platform but it must be publicly visible for our testing.
 
 Start this project on Repl.it using <a href="https://repl.it/github/freeCodeCamp/boilerplate-project-american-british-english-translator">this link</a> or clone <a href='https://github.com/freeCodeCamp/boilerplate-project-american-british-english-translator/'>this repository</a> on GitHub! If you use Repl.it, remember to save the link to your project somewhere safe!
+
+<h3>Instructions</h3>
+
+
```
Comment (moT01) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/american-british-translator.md:17]:
```suggestion
```

I feel like we don't need this

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/american-british-translator.md [hunk 3]
```diff
@@ -11,11 +11,21 @@ Build a full stack JavaScript app that is functionally similar to this: <a href=
...
 </section>
 
 ## Instructions
 <section id='instructions'>
 
+- All logic can go into `/components/translator.js`
+- Create all of the unit/functional tests in `tests/1_unit-tests.js` and `tests/2_functional-tests.js`
+- See the JavaScript files in `/components` for the different spelling and terms your application should translate
+- To run the tests on Repl.it, set NODE_ENV to test without quotes in the .env file
```
Comment (moT01) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/american-british-translator.md:26]:
```suggestion
- To run the tests on Repl.it, set `NODE_ENV` to `test` without quotes in the `.env` file
```

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/american-british-translator.md [hunk 4]
```diff
@@ -11,11 +11,21 @@ Build a full stack JavaScript app that is functionally similar to this: <a href=
...
+- To run the tests in the console, use the command npm run test. To open the Repl.it console, press Ctrl+Shift+P (Cmd if on a Mac) and type "open shell"
```
Comment (moT01) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/american-british-translator.md:27]:
```suggestion
- To run the tests in the console, use the command `npm run test`. To open the Repl.it console, press Ctrl+Shift+P (Cmd if on a Mac) and type "open shell"
```

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/american-british-translator.md [hunk 5]
```diff
@@ -28,28 +38,225 @@ tests:
...
-      If there is no text in the text area, append the message "Error: No text to translate." to the <code>error-msg</code> <code>div</code> so the text appears in red.
-    testString: ''
-  - text: I can press the "Clear Input" button to remove all text from the text area and the <code>translated-sentence</code> <code>div</code>.
-    testString: ''
-  - text: All 20 unit tests are complete and passing. See <code>/tests/1_unit-tests.js</code> for the sentences you should write tests for.
-    testString: ''
-  - text: All 4 functional tests are complete and passing. See <code>/tests/2_functional-tests.js</code> for the functionality you should write tests for.
-    testString: ''
+
+  - text: I can <code>POST</code> to <code>/api/translate</code> with a body containing <code>text</code> with the text to translate and <code>locale</code> with either <code>american-to-british</code> or <code>british-to-american</code>, The returned object should contain the submitted <code>text</code> and <code>translation</code> with the translated text.
```
Comment (moT01) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/american-british-translator.md:42]:
```suggestion
  - text: You can <code>POST</code> to <code>/api/translate</code> with a body containing <code>text</code> with the text to translate and <code>locale</code> with either <code>american-to-british</code> or <code>british-to-american</code>. The returned object should contain the submitted <code>text</code> and <code>translation</code> with the translated text.  
```

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/american-british-translator.md [hunk 6]
```diff
@@ -28,28 +38,225 @@ tests:
...
+        let parsed = await data.json();
+        assert.isObject(parsed);
+        assert.property(parsed, 'text');
+        assert.property(parsed, 'translation');
+        assert.deepEqual(parsed, output);
+      } catch (err) {
+        throw new Error(err.responseText || err.message);
+      }
+    }"
+  - text: The <code>/api/translate</code> route should also handle the way titles/honorifics are abbreviated in American and British English. For example, Doctor Wright is abbreviated as "Dr Wright" in British English and "Dr. Wright" in American English. See <code>/public/american-to-british-titles.js</code> for the different titles your application should handle.
```
Comment (moT01) [inline curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/american-british-translator.md:92]: I worry that when we say "Doctor Wright is abbreviated..." that users will translate "doctor" to `Dr/Dr.` - Is that something the app is supposed to do? Similar issue with the test above this one. Perhaps a slight reword... Maybe, "For example, `Dr Wright` in British English would be `Dr. Wright` in American English.

Perhaps it's not a big deal if they do or don't translate that, not sure if there's any tests for it.

**Review by moT01 — CHANGES_REQUESTED** [review]
I left a few suggestions, most of them are pretty minor.

Can you change the first test text to `You should provide your own project, not the example URL.`

Can you also update the description to.
```
Build a full stack JavaScript app that is functionally similar to this: <a href='https://american-british-translator.freecodecamp.rocks/' target='_blank'>https://american-british-translator.freecodecamp.rocks/</a>. Working on this project will involve you writing your code using one of the following methods:

- Clone <a href='https://github.com/freeCodeCamp/boilerplate-project-american-british-english-translator/' target='_blank'>this GitHub repo</a> and complete your project locally.
- Use <a href='https://repl.it/github/freeCodeCamp/boilerplate-project-american-british-english-translator' target='_blank'>our repl.it starter project</a> to complete your project.
- Use a site builder of your choice to complete the project. Be sure to incorporate all the files from our GitHub repo.

When you are done, make sure a working demo of your project is hosted somewhere public. Then submit the URL to it in the `Solution Link` field. Optionally, also submit a link to your project's source code in the `GitHub Link` field.
```

If you feel like it, pretty sure you can use backticks instead of all those `<code>` tags in the test text - I don't mind either way.

This is just a review of the language, I still need to check out the tests. It's looking good though 🎉 

Edit: I am going through this now. Should we add something to those instructions about the route. "Complete the route in `/routes/api.js`"?

I think the tests that check the output should allow for single or double quotes. Something like this maybe...
`translation: '<span class=('|")highlight\1>Dr</span> Grosh will see you now.'`
I didn't play with it to find out how to make it work. I'm just adding things as I find them here.

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/american-british-translator.md (hunk 2 above)
Comment (SaintPeter) [reply curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/american-british-translator.md:17]: This is part of the Project Markdown Boilerplate, as near as I can tell.  I didn't put it there, it was already there.  I'm disinclined to remove it unless that's the standard.

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/american-british-translator.md (hunk 6 above)
Comment (SaintPeter) [reply curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/american-british-translator.md:92]: I didn't have any issues with this, considering that it's followed by the reference to the `american-to-british-titles.js` which clearly spells out the translations.  We don't have any tests for "Doctor" that I am aware of, so it seems like a moot point.

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/american-british-translator.md (hunk 2 above)
Comment (SaintPeter) [reply curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/american-british-translator.md]: I just realized it was a duplicated header.  I think it was a "cut and paste error".

File: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/american-british-translator.md (hunk 6 above)
Comment (moT01) [reply curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/american-british-translator.md:88]: Yea, I'm fine with it then.

**Review by moT01 — DISMISSED** [review]
Did you have any thoughts on this comment of mine @SaintPeter @nhcarrigan... ?
> I think the tests that check the output should allow for single or double quotes. Something like this maybe...
translation: '<span class=('|")highlight\1>Dr</span> Grosh will see you now.'

The text does say specifically to use double quotes...
`Wrap any translated spelling or terms with <span class="highlight">...</span> tags so they appear in green.`

so I suppose we are probably fine - perhaps that should be emphasized with `Be sure to use double quotes` or something. But I guess if we are not going to change it to allow single or double quotes, then I probably shouldn't hold this PR up for an extra line of text.

I looked at all three PR's and, other than that one concern, I think they all look good to go - I am thinking that it's not worth it to change that at this point. I would rather get these out there, and if problems arise, we can change it later. So I am just going to approve them. Nice work on these @SaintPeter and @nhcarrigan 🎉

**Review by RandellDawson — CHANGES_REQUESTED** [review]
@SaintPeter I just have a couple suggestions and then I can approve it and let @moT01 look at it once more before merging.

**Review by RandellDawson — APPROVED** [review]

**Comment by RandellDawson** [comment]
@Sky020 @moT01 I am now fine with this PR and the associated boilerplate and example project PRs. If you both approve, feel free to merge this one.

**Review by moT01 — APPROVED** [review]
LGTM 🎉 Thanks again @SaintPeter and @nhcarrigan 🎉

**Comment by SainiKhushi** [comment]
Kindly share your codework for American British translator please

**Pull request merged**